README.md
# ProjectUtils

这个工程是项目中可能会用到的各种 Utils 集合

是为了方便 Android 开发者,在使用一些日常 utils 的时候,方便寻找和使用.


## 使用方法
* 关于项目中的 application
    * 如果项目中并未自己实现 Application ，则将 AndroidManifest.xml 中的 application 节点中的 android:name="" 写为 android:name="com.sunhz.projectutils.base.BaseApplication"
    * 如果项目中已经创建了自己的 application 但未继承其他 SDK 中 application，则直接将 com.sunhz.projectutils.base.BaseApplication 写为自己的 application 的父类。
    * 在 com.sunhz.projectutils.base.BaseApplication 中的 onCreate 方法中，调用了用来初始化 Constant 的方法。具体说明请看下面。

* 关于 com.sunhz.projectutils.Constant 的说明
    * Constant 类中有三个值。
        * Encryption.AES_UTILS_CLIENT_KEY 这个值是使用 AES 加解密工具时的 key 。如果需要使用该功能，则需要填写十六位 key，若不需要使用，则不需要填写，默认为 ""。
        * TimeInApplication.NET_TIMEOUT 这个值是网络请求超时时间，网络请求超时分为两种，一个是请求超时，一个是响应超时，两种请求超时的时间在这里设置成同样的。默认为 10s。
        * TimeInApplication.CACHE_FAIL_TIME 这个值是缓存到本地的数据失效时间，默认为一天。
    * 设置方式有两种。1，继承com.sunhz.projectutils.base.BaseApplication。2，手动在自己的 application 中调用 Constant.initConstant(aes key, net time out, cache fail time) 或者 Constant.initNetTimeOutAndCacheFailTime（net time out, cache fail time) 方法。
    * 若不继承 com.sunhz.projectutils.base.BaseApplication 或不在自己的 application 中调用 Constant.initConstant(aes key, net time out, cache fail time) 或者 Constant.initNetTimeOutAndCacheFailTime（net time out, cache fail time) 方法，则以默认值为准。

* 关于项目中的 cache 功能
    * 如果需要使用缓存功能，则需要在 application 中的 onCreate 方法中，调用 CacheUtils.getInstance(Context).init()。

* 关于判断第一次启动 app 的工具类
    * 如果需要是用判断 app 是否是第一次启动，则需要在 application 中的 onCreate 方法中，调用  FirstStartUtils.init(Context)。

* 关于开启记录软件崩溃日志功能
    * 如果需要使用记录崩溃日志功能，则需要在 application 中的 onCreate 方法中，调用 CrashHandler.getInstance(Context).init()。

* 关于使用 logUtils 的工具类
    * 如果项目需要使用 logUtils，则需要在 application 中的 onCreate 方法中，调用 Logger.init().setMethodCount(3).setLogLevel(LogLevel.FULL)。
    * 如果暂时需要关闭，则覆盖上面的代码，在 application 中的 onCreate 方法中，调用 Logger.init().setLogLevel(LogLevel.NONE)。

### 使用方法整合版
```java

    // 开启缓存功能
    CacheUtils.getInstance(mContext).init();

    // 开启记录 app 第一次打开功能
    FirstStartUtils.init(Context);

    // 设置项目的 AES 加密方式的 key，设置网络超时时间（请求超时，响应超时为同一时间），设置缓存失效时间
    Constant.initConstant("xxxxxxxxxxxxxxxx", 10000, BaseApplication.A_DAY);

    // 如果项目不需要使用 AES 加密工具，就不需要设置 AES 加密方式的 key，可以忽略上面方法， 直接调用下面方法。
    Constant.initNetTimeOutAndCacheFailTime(1000,BaseApplication.A_DAY);

    // Fix ClipboardUIManager memory leak
    FixClipboardUIManagerLeak.fixClipboardUIManagerLeak(this);

    if (!BuildConfig.DEBUG) { // 正式环境
        // 开启记录崩溃日志功能
        CrashHandler.getInstance(mContext).init();

        // 关闭 logUtils
        Logger.init().setLogLevel(LogLevel.NONE);
    } else { // 开发环境
        // 开启 logUtils
        Logger.init().setMethodCount(3).setLogLevel(LogLevel.FULL);
    }


```

## 目录结构说明
### 缓存目录结构说明
```
// 有SD卡, 并可以使用的情况
- /sdcard
    - android
        - package name (程序包名)
            - files
                - cache (缓存目录)
                    - cacheData (缓存数据)
                    - cacheImage (缓存图片)
                    - cacheOther (缓存其他东西)
                        - log (程序异常崩溃日志)

// 无SD卡,或SD卡不可使用的情况
- /data/data
    - package name (程序包名)
        - cache (缓存目录)
            - cacheData (缓存数据)
            - cacheImage (缓存图片)
            - cacheOther (缓存其他东西)
                - log (程序异常崩溃日志)
```

## 问题汇总

>    项目中的 LogUtils 类是封装使用了 orhanobut 的 logger 工具库。
>    logger 这个项目会将 Log 从新排版，并将 log 用表格的方式展示出来。

>    但在使用过程中需要注意，如果使用 ProjectUtils 中的 LogUtils.e(exception)，这是打印出来的，并没有表格。
>    现象原因：ProjectUtils 中的 LogUtils.e(exception) 里面只是调用了系统的 exception.printStackTrace() 方法，由系统来打印错误信息。

>    背景原因：因为用 logger 库中的 logger.e(Throwable throwable, String message, Object... args) 方法打印出来的异常信息，并不会把当前出现异常代码的行号告诉我们，只能告诉我们打印这个异常信息的位置。
>    因此对我们解决问题并没有特别大的帮助，所以将错误一场打印换为 JAVA JDK 中的方式。

>    不过用这个 logger 的好处是，我们在打印普通信息 log 的时候也能够方便的找到打印log的代码位置，并且用漂亮的表格将我们需要的信息打印到控制台上。
>    并且 ProjectUtils 使用 LogUtils 类封装了 logger 后已经对 LogUtils 进行了统一管理。方便我们在 release 和 debug 版本间任意切换。避免了出现忘记了删除 log 语句的问题。


## 引用到的开源项目说明

其中使用了一些其他的开源项目：

[com.android.support-v4_21](https://developer.android.com/tools/support-library/features.html "com.android.support:support-v4:21.0.0")

[Gson_2.3.1](https://code.google.com/p/google-gson/ "com.google.code.gson:gson:2.3.1")

[Volley_1.0.15](https://developer.android.com/training/volley/index.html "com.mcxiaoke.volley:library:1.0.15")

[logger_1.10](https://github.com/orhanobut/logger "com.orhanobut:logger:1.10")

所以若用到以上功能，上述引用便可不用再次添加。




###License

```
Copyright (c) 2015 , 给立乐 Spencer ChinaSunHZ (www.spencer-dev.com).

Licensed under the Apache License, Version 2.0 (the "License”);
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


release-notes.md
#ProjectUtils

## 最新版本 : version 0.1


## release 0.1 log
* Base：提供 BaseActivity，BaseFragment，BaseApplication，Base。 以方便管理整个项目的 activity 和 fragment。

* AsyncTask：提供 AsyncTaskExpand 。以简化对 AsyncTask 的串行并行管理操作。

* Cache：提供 CacheType 、CacheUtils。以简化对 Cache 类型的管理，以及对 Cache 的读取写入操作。

* Device：提供 DeviceUtils。以简化对 Device Info  的获取。

* Encryption：提供 AESUtils、Base64Utils、MD5EncryptionUtils。以简化对 AES、Base64、MD5 的加解密操作。

* File：提供 FileUtils、SDCardUtils、SharePreferenceUtils。
	* FileUtils 用来简化对文件的读取写入和获取文件信息时的操作。
	* SDCardUtils 用来简化对 SDCard 相关操作，如判断是否存在，是否可写等。
	* SharePreferenceUtils 用来简化对 SharePreference 的日常操作需求。

* FirstStart：提供 FirstStartUtils 。以简化对应用的第一次启动的各种判断。

* Http：提供 DownloadImageFromUrl、GsonRequest、HttpClientUtils、UrlParamsString。
	* DownloadImageFromUrl 用来简化对单张图片的下载操作。
	* GsonRequest 对请求的结果进行了封装，将结果以泛型对象形式返回，以方便调用者使用。
	* HttpClientUtils 将常用的正常请求封装起来，以简化我们对正常请求使用的需求，此工具类的方法，需要在子线程中使用。
	* UrlParamsString 是对 Http 请求中的 get 请求的参数进行封装，以此来简化平时使用时对参数的字符串拼装。

* Image：提供 ImageUtils。用来简化对图片的操作，如压缩图片，bitmap 转 drawable，drawable 转 bitmap 等常用操作。

* Intent：提供 IntentUtils。用来简化平时对 activity 时所需要操作，并提供了传参以及不传参的跳转方式。

* Json：提供了 GsonUtils。用来简化对 json 转 obj 的操作。并且提供了方便的 json 转 obj 、json 转 list、obj 转 json 的操作。

* Log：提供了 LogUtils。以方便对 log 的管理，异常本地保存，以及对 log 的展示美化等。

* Media：提供了 MediaUtils。以方便将图片、视频、音频插入以及删除系统媒体库。

* NetWork：提供了 CheckInternetUtils、NetAuthorityEnum。以方便对网络的检查和网络类型的判断。

* Package：提供了 PackageUtils。以方便对软件的安装以及卸载等其他操作的调用。

* Toast：提供了 ToastUtils。提供了在子线程，主线程中显示 Toast 的便捷操作。

* View：提供了 DensityUtils、KeyboardStatusUtils、StatusBarUtils。
	* DensityUtils：用来获取屏幕信息，如分辨率，长宽，px 转 dp，dp 转 px，sp 转 px，px 转 dp 等操作。
	* KeyboardStatusUtils：用来监听软键盘是否弹出、收回，并提供相应 callBack 以方便使用。
	* StatusBarUtils：用来获取状态栏高度，选择是否隐藏以及显示状态栏的便捷操作。
	* CountDownButtonHelper：快速完成 button 的倒计时操作。

* ActivityManager：提供管理 activity 以及 closeAll activity 的便捷方法。

* Constant：存放常量，使用方法请查看 [项目说明](https://github.com/ChinaSunHZ/ProjectUtils/blob/master/README.md)

* CrashHandler：提供了将报错信息保存下来的方法。

* SoftInputUtils：提供了软键盘的开关方法。

* FixInputMethodManagerLeak：修复某些手机的 InputMethodManager 造成的内存泄露问题。

* FixClipboardUIManagerLeak：修复某些手机的 ClipboardUIManager 造成的内存泄漏问题。

* BrightnessUtils：快速设置、还原屏幕亮度。

app/src/main/java/com/sunhz/projectutils/Constant.java
/**
 * Copyright (c) 2015, Spencer , ChinaSunHZ (www.spencer-dev.com).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.sunhz.projectutils;

/**
 * Constant pool
 * Created by Spencer (www.spencer-dev.com) on 15/2/3.
 */
public final class Constant {


    /**
     * Encryption constant
     */
    public static final class Encryption {
        /**
         * AES_UTILS_CLIENT_KEY is AES encoding  Key, Key length is 16
         */
        public static String AES_UTILS_CLIENT_KEY = "";
    }


    /**
     * Program, the time control
     */
    public static final class TimeInApplication {
        /**
         * Network timeout, default 3 seconds
         */
        public static int NET_TIMEOUT = 10000;

        /**
         * Cache invalidation time, the default one day
         */
        public static int CACHE_FAIL_TIME = 1000 * 60 * 60 * 24;
    }

    /**
     * Init project utils base constant.
     *
     * @param AESUtilsClientKey aes key , default value is "".
     * @param netTimeOut        net time out , default value is 10s.
     * @param cacheFailTime     cache fail time , default value is a day.
     */
    public static void initConstant(String AESUtilsClientKey, int netTimeOut, int cacheFailTime) {
        Encryption.AES_UTILS_CLIENT_KEY = AESUtilsClientKey;
        TimeInApplication.NET_TIMEOUT = netTimeOut;
        TimeInApplication.CACHE_FAIL_TIME = cacheFailTime;
    }

    /**
     * Init project utils net time out and cache fail time.
     *
     * @param netTimeOut    net time out , default value is 10s.
     * @param cacheFailTime cache fail time, default value is a day.
     */
    public static void initNetTimeOutAndCacheFailTime(int netTimeOut, int cacheFailTime) {
        Encryption.AES_UTILS_CLIENT_KEY = "";
        TimeInApplication.NET_TIMEOUT = netTimeOut;
        TimeInApplication.CACHE_FAIL_TIME = cacheFailTime;
    }
}